### ISSUE TITLE
Auto-generate documentation from code.

### ISSUE THREAD

**Issue by jason-bragg** [dotnet/orleans#88]
Auto-generated documents can help provide more complete and up-to-date documentation with little effort.

I'm thinking first steps might be.
1) Select code generator - Doxygen?
2) Create basic configurations and script to generate reference manual like documentation.

Then we can refine the comments in code to improve the generated documents.

When generated documentation is of sufficient quality we can:
3) Incorporate document generation into the builds.
4) Add automation to publish documents online whenever repository is updated.

Thoughts?

**Comment by ghuntley** [comment]
Auto-generated documentation is almost worthless. At best it’s a slightly
    improved version of simply browsing through the source, but most of the time
    it’s easier just to read the source than to navigate the bullshit that these
    autodoc tools produce. About the only thing auto-generated documentation is good
    for is filling printed pages when contracts dictate delivery of a certain number
    of pages of documentation. I feel a particularly deep form of rage every time I
    click on a “documentation” link and see auto-generated documentation.

    There’s no substitute for documentation written, organized, and edited by hand.

    I’ll even go further and say that auto-generated documentation is worse than
    useless: it lets maintainers fool themselves into thinking they have
    documentation, thus putting off actually writing good reference by hand. If you
    don’t have documentation just admit to it. Maybe a volunteer will offer to write
    some! But don’t lie and give me that auto-documentation crap.

http://jacobian.org/writing/what-to-write/
http://jacobian.org/writing/technical-style/
http://jacobian.org/writing/editors/

Create a docs/ folder using http://www.mkdocs.org/ and hook it into http://www.readthedocs.org

Example see:

https://github.com/ghuntley/ReactiveUI/blob/readthedocs/mkdocs.yml
https://github.com/ghuntley/ReactiveUI/tree/readthedocs/docs
http://docs.reactiveui.net which is autobuilt from checkings to docs/* by RTD.

**Comment by jthelin-microsoft** [comment]
There is already lots of human-written documentation in the wiki which is [at least is intended to be] along the lines that @ghuntley describes. 
https://github.com/dotnet/orleans/wiki

While we all try to be professional, at the end of the day we are all still Developers not Copy Editors, and I don't think any of us gets up in the morning with our first thought being "Hmm, today I feel like writing some documentation!" ;)

So, if there are specific topics that you feel are NOT already covered well in the wiki, then please flag them up as enhancement request Issues on GitHub. We will probably grumble a little (hey, we are human too!) but then will try to write up something "decent" for community review / feedback.

Also, if anyone is exploring specific areas of the code base and wants to provide any summaries / overviews / knowledge capture for those parts, then we are very open to those being incorporated into the wiki too.

At the end of the day, these are not "our" docs, they are "Our" docs belonging to the Community.

-----

Having said that, there is also still a place for doc-comments for APIs, and in general it is good discipline for developers to write explanations for method / class purpose and meaning of each parameter, at least for the public parts of the APIs. 

I think @jason-bragg query is about people's experiences and suggestions for specific *good* tools which help to auto-generate API docs to HTML as API's change / evolve, and which integrate well with GitHub / Jenkins-CI.

The original core team for Orleans tried to be fairly pragmatic, but we also did aim to follow the min-bar of having "decent" code-docs for all public API classes & methods. 
I strongly suggest that this practice is continued by all contributors going forward.

**Comment by gabikliot** [comment]
I agree with @jthelin-microsoft .
Regarding the documentation for specific areas of the code base, we started this section: https://github.com/dotnet/orleans/wiki/Runtime-Implementation-Details and plan to extend it with description of more internal components.

**Comment by sergeybykov** [comment]
Also, @richorama  has written a bunch of blog posts on various important topics. I'm not sure what's the best way to integrate them. Add links to a relevant wiki page?

**Comment by richorama** [comment]
Someone should write a book...

I don't mind contributing a chapter, but I don't think I could muster the energy to write the whole thing...


-----Original Message-----

**Comment by jason-bragg** [comment]
I apologize for being unclear in my initial proposal.  By ‘reference manual’, I was referring to public API documentation.  Samples such as "My First Orleans Application" (https://github.com/dotnet/orleans/wiki/My-First-Orleans-Application), descriptions of framework behaviors like "Reentrant Grains" (https://github.com/dotnet/orleans/wiki/Reentrant-Grains) , and many other such documentation needs are, in my experience, not well met by generated documentation.  Auto-generated API reference documentation, has the advantage of being comprehensive, and up to date.  Additionally, once the initial integration work is done, it’s extremely light wait, as the cost of maintaining it is negligible, because it requires only that developers maintain clear comments in public APIs, something we should be doing anyway.

**Comment by ToddShelton** [comment]
@jason-bragg  are you talking about ///-style inline documentation or a separate API document?

**Comment by MovGP0** [comment]
When someone wants to have a Doxygen file, the person can easily adopt the build file to do so: 

     <?xml version="1.0" encoding="utf-8"?>
    <Project ... DefaultTargets="Doxygen" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
      
      <Target Name="Doxygen" Condition="Exists('bin\doxygen.exe')">
        <MakeDir Directories="$(builtdir)\docs" />
        <Exec Command="bin\doxygen.exe" />
      </Target>
    </Project>

However, I discourage to do this by default. To my experience there is not much benefit to this when the project is open source. Its way better to look at the actual code.

**Comment by veikkoeeva** [comment]
> Someone should write a book... I don't mind contributing a chapter, but I don't think I could muster the energy to write the whole thing...

Indeed. Really nice would be if it had a brief tour on some theory with regards to [actor model (agents?)](https://en.wikipedia.org/wiki/Actor_model), [CAP](https://en.wikipedia.org/wiki/CAP_theorem)(*), [FLP](https://en.wikipedia.org/wiki/Consensus_%28computer_science%29#Solvability_results_for_some_agreement_problems) and some higher level design traits. From a more engineering angle, throughput (it's interesting that embedded systems work like Orleans to guarantee throughput and then there are special purpose circuits for heavy algoritms) and latency considerations, scheduling (cooperative/preemptive), reentrancy and whatnot, design patterns pitfalls etc. Then some actual code and constructs sprinkled in so one doesn't have to try to guess how what are the actual constructs in code.

(*) I'm not sure about the right level of exposure. Perhaps something like in
* [Don't Settle for Eventual Consistency](https://queue.acm.org/detail.cfm?id=2582994)
* [Eventual Consistency Today: Limitations, Extensions, and Beyond](https://queue.acm.org/detail.cfm?id=2610533)
* and the [Jepsen Call Me Maybe series](https://aphyr.com/tags/jepsen).

**Comment by richorama** [comment]
Sounds good, I look forward to reading it!

-----Original Message-----

**Comment by gabikliot** [comment]
http://dotnet.github.io/docfx/
Looks relevant to what you wanted to do @jason-bragg .

**Comment by jdom** [comment]
We are now using DocFX, so we are closer to this goal

**Comment by ReubenBond** [comment]
Closing this until it's actionable. Perhaps a future DocFx release will make this easier.
